### ISSUE TITLE
race condition: "btrfs subvolume create failed" with concurrent image fetches

### ISSUE THREAD

**Issue by candlerb** [lxc/lxd#2739]
# Required information

 * Distribution: ubuntu
 * Distribution version: 16.04, with btrfs on /var/lib/lxd

~~~
apiextensions:
- id_map
apistatus: stable
apiversion: "1.0"
auth: trusted
environment:
  addresses:
  - 192.168.5.54:8443
  - 10.0.8.1:8443
  architectures:
  - x86_64
  - i686
  certificate: <<snip>>
  certificatefingerprint: <<snip>>
  driver: lxc
  driverversion: 2.0.5
  kernel: Linux
  kernelarchitecture: x86_64
  kernelversion: 4.4.0-53-generic
  server: lxd
  serverpid: 2218
  serverversion: 2.0.8
  storage: btrfs
  storageversion: "4.4"
config:
  core.https_address: '[::]'
  core.trust_password: true
public: false
~~~

Using btrfs for /var/lib/lxd:

~~~
# mount | grep btrfs
/dev/vda3 on /var/lib/lxd type btrfs (rw,relatime,space_cache,subvolid=5,subvol=/)
/dev/vda3 on /var/lib/lxd/shmounts type btrfs (rw,relatime,space_cache,subvolid=5,subvol=/shmounts)
~~~

# Issue description

Launching multiple containers concurrently, where the image is not in the local image store.

Sometimes, the same image is fetched multiple times concurrently, and then we get an error because the second download finds that the image already exists.

Looks like some locking is required?

Reason this is a problem: using an ansible script to create a bunch of containers with the lxd_container module, by default it hits the REST API concurrently with a bunch of simultaneous creates, and this problem occurs sometimes when the local image doesn't yet exist. (Workaround is to tweak ansible config to create them sequentially instead of in parallel).

# Steps to reproduce

Choose an image to test with, e.g. 'ubuntu:14.04'

Delete the local copy of this image if you already have it.

~~~
root@wrn-dns2:~# lxc image list
+-------+--------------+--------+---------------------------------------------+--------+----------+-------------------------------+
| ALIAS | FINGERPRINT  | PUBLIC |                 DESCRIPTION                 |  ARCH  |   SIZE   |          UPLOAD DATE          |
+-------+--------------+--------+---------------------------------------------+--------+----------+-------------------------------+
|       | 8573fa6fed57 | no     | ubuntu 14.04 LTS amd64 (release) (20161213) | x86_64 | 119.27MB | Dec 19, 2016 at 10:22am (UTC) |
+-------+--------------+--------+---------------------------------------------+--------+----------+-------------------------------+
root@wrn-dns2:~# lxc image delete 8573fa6fed57
~~~

Then:

~~~
# for i in foo{1..3}; do lxc launch ubuntu:14.04 $i & done
[1] 3571
[2] 3572
[3] 3573
root@wrn-dns2:~# Creating foo3
Creating foo1
Creating foo2
Retrieving image: 100%error: Unpack failed, exit status 2.  tar (child): /var/lib/lxd/images/8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88: Cannot open: No such file or directory
error: Previous download didn't succeed
error: btrfs subvolume create failed, subvol=/var/lib/lxd/images/8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88.btrfs, outputERROR: cannot create subvolume: File exists
Create subvolume '/var/lib/lxd/images/8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88.btrfs'


[1]   Exit 1                  lxc launch ubuntu:14.04 $i
[2]-  Exit 1                  lxc launch ubuntu:14.04 $i
[3]+  Exit 1                  lxc launch ubuntu:14.04 $i
~~~

Note 1: this doesn't happen every time. You may need to delete containers foo1/2/3, delete the image, and repeat.  But when it does happen, none of foo1/2/3 are created.

Note 2: when the problem does occur, during the "Retrieving image" stage, you can see the percentage progress figure jumping up and down; it's clearly doing two downloads at the same time.

The test machine is a VM with two virtual cores. Maybe it would be easier to reproduce with more cores, i.e. more concurrent activity?

# Information to attach

## /var/log/lxd/lxd.log

 - [X] any relevant kernel output (dmesg)

Nothing apart from usual btrfs noise:

~~~
[585509.526245] BTRFS error (device vda3): could not find root 8
[585509.528471] BTRFS error (device vda3): could not find root 8
[585509.534235] BTRFS error (device vda3): could not find root 8
~~~

 - [X] container log (lxc info NAME --show-log)

Nothing (containers not created)

 - [X] main daemon log (/var/log/lxd.log)

~~~
t=2016-12-19T10:23:15+0000 lvl=info msg="Downloading image" trigger=/1.0/operations/09195a62-5213-425e-840c-97c1fa8a55ea image=8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88 operation=09195a62-5213-425e-840c-97c1fa8a55ea alias=14.04 server=https://cloud-images.ubuntu.com/releases
t=2016-12-19T10:23:15+0000 lvl=info msg="Downloading image" operation=b3c68070-6392-426e-8069-6bc46a9299fa alias=14.04 server=https://cloud-images.ubuntu.com/releases trigger=/1.0/operations/b3c68070-6392-426e-8069-6bc46a9299fa image=8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88
t=2016-12-19T10:23:28+0000 lvl=eror msg="Previous download didn't succeed" image=8573fa6fed57ecadbb6f76f6b424533e4a8c5a562e29cee28ba47b6923aa4f88
~~~

**Comment by stgraber** [comment]
There actually is locking around image imports to avoid specifically this kind of problem, so I'm not sure exactly what's going on here...

**Comment by stgraber** [comment]
I've been unable to reproduce this even after several tries using the provided loop.

I have however noticed a very small window for a race condition in the image downloading code which I've now written a patch to fix. Hopefully that was the source of this problem and we can consider this resolved.

**Comment by candlerb** [comment]
Thank you.

I couldn't reproduce it directly on a quad-core host. However I have checked that I can still reproduce it on this somewhat resource-constrained KVM VM (*), and I can. It took 4 attempts to make it happen again:

~~~
...
root@wrn-dns2:~# lxc delete -f foo{1..3}; sync; lxc image delete 5f19596a09c1; sync; sleep 1; for i in foo{1..3}; do lxc launch ubuntu:14.04 $i & done; wait
[1] 31225
[2] 31226
[3] 31227
Creating foo2
Creating foo1
Creating foo3
Retrieving image: 100%error: Unpack failed, exit status 2.  tar (child): /var/lib/lxd/images/5f19596a09c19d4beed653de08ea438c32e2fee35a10c6963a8285cae95c0a43: Cannot open: No such file or directory
error: btrfs subvolume create failed, subvol=/var/lib/lxd/images/5f19596a09c19d4beed653de08ea438c32e2fee35a10c6963a8285cae95c0a43.btrfs, outputERROR: cannot create subvolume: File exists
Create subvolume '/var/lib/lxd/images/5f19596a09c19d4beed653de08ea438c32e2fee35a10c6963a8285cae95c0a43.btrfs'

[1]   Exit 1                  lxc launch ubuntu:14.04 $i
[2]-  Exit 1                  lxc launch ubuntu:14.04 $i
~~~

So if this patch is back-ported to stable-2.0, and ends up in 2.0.9, I should be able to confirm if the problem is fixed.

(*) 2 cores, 4GB, has 4 other containers running; one of these is the 'unifi' controller which has a mongodb database and java amongst other things.

**Comment by stgraber** [comment]
Cool, and yeah, the patch will most certainly be included in 2.0.9.

**Comment by candlerb** [comment]
With lxd 2.0.9 I have tried to reproduce 25 times unsuccessfully in the same environment, so I think problem is fixed. Thanks again!
